Refs #21231 -- Backport urllib.parse.parse_qsl() from Python 3.8.

pope1ni opened the pull request:
This backports [`urllib.parse.parse_qsl()`](https://docs.python.org/3.8/library/urllib.parse.html#urllib.parse.parse_qsl) from Python 3.8 to replace `limited_parse_qsl()`.

I think there may be a couple of projects using this undocumented function, so I don't know whether we want to mention it in backwards compatibility notes or whether we need to deprecate?

It seems that support for `max_num_fields` was also patched into Python 3.6.8 and 3.7.2 ([bpo-34866](https://bugs.python.org/issue34866)), so I don't know whether we even want the backport considering Django's policy on using the latest patch release in a series. That would run the risk that thinks break for people though. Even if we decide to keep it, we should probably use the version available in Python in preference for all versions of Python. What do you think?

jdufresne commented on django/http/request.py:
```diff
@@ -432,8 +438,16 @@ def __init__(self, query_string=None, mutable=False, encoding=None):
+            for key, value in parse_qsl(query_string, **parse_qsl_kwargs):
+                self.appendlist(key, value)
+        except ValueError as e:
+            if str(e) != 'Max number of fields exceeded':
```
This feels a bit iffy to me and I would not consider the message a stable API. What other `ValueError` could be thrown here that makes this necessary?

pope1ni replied:
When updating the backported function, there was new code added that includes an additional `ValueError`. We are not using the new `strict_parsing` flag, however, so we could just assume the that `ValueError` raised is not from the branch guarded by that boolean flag?

pope1ni commented:
Ok. I've pushed a `!fixup` that removes the fragile string checking on the `ValueError` exception raised by `urllib.parse.parse_qsl()`. This only works because we are using the default `strict_parsing=False`.

I have also added in a second `!fixup` that strips out the backport entirely. The change to add `max_num_fields` to `urllib.parse.parse_qsl()` was released in Python 3.6.8 and 3.7.2 on 24th December 2018. Django has a policy of only supporting the latest patch releases, but this may be deemed too soon - I've added the commit to provide this option should it be preferred at the discretion of one of the fellows.

pope1ni commented:
@felixxm I figured it would be too early, but worth asking. Will wait for Carlton's assessment as you've cc'd him and then remove the `!fixup` if he concurs.

Were you happy enough with the approach to `ValueError` → `TooManyFieldsSent`?

carltongibson commented:
Yep. I don’t see that we need to push here. (I.e. I’d keep the fallback for the now.)

pope1ni commented:
Ok. I've updated this to keep the backport and squashed the other fixup. Should be ready to go.

carltongibson commented on django/http/request.py:
```diff
@@ -19,11 +19,18 @@
 
 from .multipartparser import parse_header
 
+if PY38:
```
I'd probably just add a comment here with the version support details, so when we come across this again ≈4.x or so we can review.

pope1ni commented:
We should always prefer the standard library version if we have a version that supports `max_num_fields`, so I changed it to use:
```python
if (3, 6, 8) <= sys.version_info[:3] < (3, 7, 0) or (3, 7, 2) <= sys.version_info[:3]:
    from urllib.parse import parse_qsl
else:
    from django.utils.http import parse_qsl
```
I've already added a `# TODO:` on `django.utils.http.parse_qsl()` that it should be removed when 3.8 is the minimum supported version of Python.

felixxm commented on django/http/request.py:
```diff
@@ -19,11 +20,17 @@
 
 from .multipartparser import parse_header
 
+if (3, 6, 8) <= sys.version_info[:3] < (3, 7, 0) or (3, 7, 2) <= sys.version_info[:3]:
```
I'm not sure about this check. I would prefer to use `urllib.parse.parse_qsl` only on Python 3.8 which seems reasonable because you backported version from Python 3.8 to the `django.utils.http.parse_qsl`.

Alternatively, we can check `urllib.parse.parse_qsl` signature rather then Python version on the grounds that adding `max_num_fields` was a security patch so older versions may have this patch:
```
Python 3.6.3
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from urllib.parse import parse_qsl
>>> help(parse_qsl)
​
Help on function parse_qsl in module urllib.parse:
​
parse_qsl(qs, keep_blank_values=False, strict_parsing=False, encoding='utf-8', errors='replace', max_num_fields=None)
    Parse a query given as a string argument.
```
but I would prefer to use `urllib.parse.parse_qsl` only on Python 3.8.

pope1ni commented:
@felixxm I've switched back to using `PY38` in the main commit, but added a `!fixup` using `inspect.signature()` for consideration. I feel it may be worth looking forward at this point and using the implementation in Python where available. If you feel otherwise, you can drop the `!fixup`.

@carltongibson I've added another `!fixup` that pulls in the unit tests I could find in python/cpython. We didn't have this before and the backport is a direct copy-paste from Python itself, but you requested tests so that seemed the most sensible approach. Again, you can drop the `!fixup` if you have changed your mind.

carltongibson commented:
Ok, cheers Nick. Thanks. I’ll give it another look.

pope1ni commented on django/http/request.py:
```diff
@@ -455,8 +461,19 @@ def __init__(self, query_string=None, mutable=False, encoding=None):
+            for key, value in parse_qsl(query_string, **parse_qsl_kwargs):
+                self.appendlist(key, value)
+        except ValueError as e:
+            # Note that ValueError can also be raised if the strict_parsing
+            # argument to parse_qsl() is True. As that is not used by Django,
+            # we assume that the exception was raised by exceeding the value of
+            # max_num_fields instead of fragile checks of exception message
+            # strings.
```
I also added this comment based on a previous review from @jdufresne. I feel it is worth mentioning this here.

carltongibson replied:
It's OK, I ❤️ comments 😀

carltongibson commented:
> @carltongibson I've added another !fixup that pulls in the unit tests I could find in python/cpython. We didn't have this before and the backport is a direct copy-paste from Python itself, but you requested tests so that seemed the most sensible approach. Again, you can drop the !fixup if you have changed your mind.

Thanks for this @pope1ni. I think we need the tests, even if it is just copy and paste, so 👍

carltongibson commented on django/http/request.py:
```diff
@@ -21,13 +22,11 @@
+# TODO: Remove when Python 3.8 is the minimum required version for Django.
+if 'max_num_fields' not in inspect.signature(parse_qsl).parameters:
     from django.utils.http import parse_qsl
 
```
I guess I'm happy with the `inspect` usage. (Is it too cute?) I'd probably add a comment explaining the why. 
I'll give it a run with an old version to double check... 

I'd rather keep the compatibility import as a single `if...else...` block. Having the main import above and then replacing here isn't as obvious as it could be, for me.

pope1ni replied:
It cannot be an `if ... else ...` block as `inspect.signature()` needs to look at the signature of `parse_qsl()` from `urllib.parse` which means it needs to be imported first. Hence this override if required approach. I also didn't want to move the import from the standard library just above the `if` otherwise we'd just be hacking around with exclusion comments for `isort`.

pope1ni replied:
Comment added.

carltongibson replied:
Grrr... I think I probably preferred the `if PY38` version then. (Let me confer with Mariusz.)

pope1ni commented:
> I couldn't find any projects on GitHub using it but, a deprecation shim would be quite light-weight. (It can just map to parse_qsl I think.)

I've added another `!fixup` with a backward incompatibility note pointing to the function in the standard library. I think a whole shim for this which would need a function (due to the different argument name), a deprecation warning and extra tests for both parts of that is a bit excessive for an undocumented function removal for which the resolution is straightforward.

The only reason for someone using this is if they're trying to parse raw query strings themselves and they knew that this function provided them with a security fix that the standard library version didn't at the time. I think this is very few people, if any.

pope1ni commented:
@carltongibson Rebased. Is there anything else you'd like me to do here?

pope1ni commented:
Rebased targeting 3.2.

pope1ni commented:
@felixxm I think @carltongibson wanted your opinion on how we do the compatibility import: https://github.com/django/django/pull/12017#discussion_r409514541.

carltongibson commented:
Hey @pope1ni. Sorry for the hold-up here. You got bumped because of the feature freeze. You're still on my list. 😉

pope1ni commented:
@carltongibson Is this one still on your list? :innocent:

carltongibson commented:
Hey @pope1ni — Yes, it is. Thank you for your patience. It's currently on the _look at in the lull after all the release blockers from 3.1 final turn up_ list. 😀

> The only reason for someone using this is if they're trying to parse raw query strings themselves and they knew that this function provided them with a security fix that the standard library version didn't at the time. I think this is very few people, if any.

This point is noted. Main issue here (from the review POV) is having the mental space to think through whether we can just note it as a breaking change (which _maybe_ we could, yes...)

carltongibson reviewed (approved):
Hey @pope1ni. Thank you for your patience here. 

I'm inclined to take this as is: 

* We have the fallback. (The `inspect` usage is fine.)
* The fallback is tested.
* The breaking change note is sufficient. 

Fancy squashing down? 

I'll double check that @felixxm doesn't hate it, but LGTM. 🥇

felixxm commented on django/http/request.py:
```diff
@@ -19,11 +20,20 @@
 
 from .multipartparser import parse_header
 
+# TODO: Remove when Python 3.8 is the minimum required version for Django.
```
`# TODO: Remove when dropping support for PY37.`

felixxm reviewed (commented):
@pope1ni Thanks :+1: 

I pushed minor edits.

pope1ni commented:
Thanks!

timgraham commented on django/http/request.py:
```diff
@@ -19,11 +19,20 @@
+# detect whether the max_num_fields argument is available as this security fix
+# was backported to Python 3.6.8 and 3.7.2, and may also have been applied by
+# downstream package maintainers to other versions in their repositories.
+if func_supports_parameter(parse_qsl, 'max_num_fields'):
```
A small oversight I noticed in an old Python 3.7.0 virtualenv: https://github.com/django/django/pull/13393
